block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -1034,6 +1082,121 @@ object BlockDagFileStorage {
             }
     } yield LMDBStore[F](env, dbi)
 
+  private def readBlockHashesByDeploy[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F]
+  ): F[List[(DeployId, BlockHash)]] = {
+    def readRec(
+        result: List[(DeployId, BlockHash)]
+    ): F[List[(DeployId, BlockHash)]] =
+      for {
+        deployIdSizeOpt <- randomAccessIO.readInt
+        result <- deployIdSizeOpt match {
+                   case Some(deployIdSize) =>
+                     val blockHashBytes = Array.ofDim[Byte](BlockHash.Length)
+                     val deployIdBytes  = Array.ofDim[Byte](deployIdSize)
+                     for {
```

Comment:
This can be replaced by:
```
randomAccessIO.readFully(deployIdBytes)
.product(randomAccessIO.readFully(blockHashBytes)))
.map{case (readDeployId, readBlockHash) => ...}
.getOrElse(Sync[F].raiseError(...))
```

Reply:
I was wrong here. My suggestion didn't improve much.

Anyway. I still have a problem with this `readBlockHashesByDeploy` function because I find it quite unreadable.

So I played a little bit with the code and I got to this:

```scala
  private def readBlockHashesByDeploy[F[_]: Sync](
      randomAccessIO: RandomAccessIO[F]
  ): F[List[(DeployId, BlockHash)]] = {
    def readDeployBytes(deployIdSize : Int) : F[(DeployId, BlockHash)] = {
      val blockHashBytes = Array.ofDim[Byte](BlockHash.Length)
      val deployIdBytes = Array.ofDim[Byte](deployIdSize)

      for {
        deployIdResult <- randomAccessIO.readFully(deployIdBytes)
        blockHashBytesResult <- randomAccessIO.readFully(blockHashBytes)
      } yield
        deployIdResult.product(blockHashBytesResult)
          .getOrElse(Sync[F].raiseError[(DeployId, BlockHash)](
            BlockHashesByDeployLogIsCorrupted
          ))
    }

    Stream.continually(randomAccessIO.readInt)
      .sequence
      .flatMap( sizes =>
        sizes.takeWhile(_.isDefined)
          .traverse(readDeployBytes(_))
      )
      .map(_.toList)
  }
```

Note: This compiles but I haven't actually tested it.

Reply:
After playing some more with the code I have arrived at this:
```scala
  private def readBlockHashesByDeploy[F[_]: Sync](
      randomAccessIO: RandomAccessIO[F]
  ): F[List[(DeployId, BlockHash)]] = {
    def readDeployBytes(deployIdSize: Int): F[(DeployId, BlockHash)] =
      for {
        deployIdResult       <- randomAccessIO.readByteString(deployIdSize)
        blockHashBytesResult <- randomAccessIO.readByteString(BlockHash.Length)
        readResult           = deployIdResult.product(blockHashBytesResult)
        result               <- readResult.liftTo[F](BlockHashesByDeployLogIsCorrupted)
      } yield result

    StreamT
      .continually(randomAccessIO.readInt)
      .takeWhile(_.isDefined)
      .mapF {
        case Some(deployIdSize) => readDeployBytes(deployIdSize)
        case None               => Sync[F].raiseError[(DeployId, BlockHash)](new IllegalStateException())
      }
      .toList
  }
```

I have implemented `RandomAccessIO.readByteString` to simplify the logic here and also had to replace `Stream` with `StreamT` since calling `Stream.continually(randomAccessIO.readInt).sequence` hangs the runtime (seems like cats eagerly evaluates infinite stream while sequencing for some reason). Hence I also had to extend `StreamT` with `StreamT.continually` and `StreamT.mapF` - check them out.